src/cargo/ops/cargo_test.rs

```diff
@@ -84,38 +90,53 @@ pub fn run_tests(manifest_path: &Path,
         try!(config.shell().verbose(|shell| {
             shell.status("Running", p.to_string())
         }));
-        match ExecEngine::exec(&mut ProcessEngine, p) {
-            Ok(()) => {}
-            Err(e) => return Ok(Some(e)),
+        if let Err(e) = ExecEngine::exec(&mut ProcessEngine, p) {
+            errors.push(Box::new(e));
+            if !options.no_fail_fast {
+                break
+            }
         }
     }
+    if errors.is_empty() {
+        Ok(None)
+    } else if errors.len() == 1 {
+        Ok(Some(errors.pop().unwrap()))
+    } else {
+        let err = internal("Multiple tests failed");
```

Comment:
This is losing a lot of contextual information (e.g. why anything failed), and I don't think that it should all be discarded entirely. The error messages at least should be concatenated together to attempt to show something.